src/async-action.ts

```diff
@@ -196,13 +196,92 @@ export function asyncAction(arg1: any, arg2?: any): any {
     return createAsyncActionGenerator(name, generator)
 }
 
+export function asyncActionWithCancel(
+    callback: Function,
+    type?: string
+): Function
+
+export function asyncActionWithCancel(callback: Function, type: string): any {
+    return function wrapActionWithCancel(arg1: any, arg2?: any) {
+        // decorator
+        if (typeof arguments[1] === "string") {
+            const name = arguments[1]
+            const descriptor: PropertyDescriptor = arguments[2]
+            if (descriptor && descriptor.value) {
+                return Object.assign({}, descriptor, {
+                    value: createAsyncActionGenerator(name, descriptor.value, callback, type)
+                })
+            } else {
+                return Object.assign({}, descriptor, {
+                    set(v: any) {
+                        Object.defineProperty(this, name, {
+                            ...descriptor,
+                            value: asyncAction(name, v)
+                        })
+                    }
+                })
+            }
+        }
+
+        // direct invocation
+        const generator = typeof arg1 === "string" ? arg2 : arg1
+        const name = typeof arg1 === "string" ? arg1 : generator.name || "<unnamed async action>"
+
+        invariant(
+            typeof generator === "function",
+            "asyncAction expects function as first arg, got: " + generator
+        )
+
+        return createAsyncActionGenerator(name, generator, callback, type)
+    }
+}
+
 let generatorId = 0
+const TAKE_LATEST = 'takeLatest'
+const TAKE_EVERY = 'takeEvery'
+const globalActionsWithCancel:{[key: string]: Array<Function>} = {}
 
-export function createAsyncActionGenerator(name: string, generator: Function) {
+function addAction(name: string) {
+    if (!globalActionsWithCancel[name]) {
```

Comment:
Since this is stored in a global store, this means that this won't work if either there are to classes with a generator with the same name, or even when there are instances of the same class? As they would end all up under the same name?

Reply:
Thanks for review! Glad to share the idea which is not so strict.

0. Yes, you are right.
1. In my use case, a class with a unique name is used for creating a singleton,  so the name of the canceler will be almost unique. 
2. There are some solutions for the problem, such as: 
>*  Be composed of the class name and the method name since the method is unique in a class, or add a global counter. 
> *  Be force to input the name for **asyncActionWithCancler**